### PR TITLE
FSM-based order analysis

### PR DESCRIPTION
**Pull request by KuechA**
Analysis module to evaluate whether the order of method calls on an object fulfills the one specified by a finite state machine (more precisely a DFA).

### CODE REVIEW

File: cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/BaseOpEdge.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.analysis.fsm
+
+class BaseOpEdge(val op: String, val base: String?, val nextState: State) {
```
Comment (oxisto) [inline cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/BaseOpEdge.kt:28]: A description here would be nice. I am not quite sure just from the name what exactly this class does

File: cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFA.kt [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+    /**
+     * Creates an edge between two nodes with a given label (operator and optional base).
+     *
+     * It checks if [from] already has an outgoing edge with the same [base] and [op] but to another
+     * target node. If so, it does not add the edge end returns `false`, otherwise it creates the
+     * edge and returns `true`.
+     */
+    fun addEdge(from: State, to: State, op: String, base: String?): Boolean {
+        if (!states.contains(from)) {
```
Comment (oxisto) [inline cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFA.kt:49]: Shouldn't the `Set` already take care of this?

File: cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFA.kt [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        states.add(newState)
+        stateCounter++
+        return newState
+    }
+
+    /**
+     * Generates the string representing this FSM in DOT format. This allows a simple visualization
+     * of the resulting automaton.
+     */
+    override fun toString(): String {
```
Comment (oxisto) [inline cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFA.kt:79]: is it feasible to have this as the standard `toString`? or should we have something like a `toDotString`?

File: cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+ * the following inputs:
+ * - [consideredBases]: A set of the IDs of nodes (typically the [VariableDeclaration]) which are
+ * considered.
+ * - [nodesToOp]: A mapping between CPG nodes and their operators used by the respective edges in
+ * the DFA. Currently, we only consider [CallExpression]s.
+ * - [thisPositionOfNode]: If a non-object oriented language was used, this is a map from CPG nodes
+ * (i.e., the [CallExpression]) to the argument position serving as base of the operation.
+ */
+open class DFAOrderEvaluator(
+    var consideredBases: Set<Long?>,
```
Comment (oxisto) [inline cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt:50]: Wouldn`t `Set<Long>` be easier? To reduce unnecessary null checks

File: cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+    private val nodeIDtoEOGPathSet = mutableMapOf<Long?, MutableSet<String>>()
+    private val log: Logger = LoggerFactory.getLogger(DFAOrderEvaluator::class.java)
+
+    /**
+     * Contains the functionality which is executed if the DFA does not contain a suitable
+     * transition for the given [node]. The evaluation ensures that the [node] is relevant, i.e.,
+     * its operator is considered by the DFA and its base is subject to analysis. This means that
+     * the order is broken at [node].
+     */
+    open fun actionMissingTransitionForNode(node: Node, fsm: DFA?) {
```
Comment (oxisto) [inline cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt:63]: Make `fsm` non-nullable?

File: cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+import de.fraunhofer.aisec.cpg.passes.astParent
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+/**
+ * This class uses a [DFA] to evaluate if the order of statements in the CPG is correct. It needs
+ * the following inputs:
+ * - [consideredBases]: A set of the IDs of nodes (typically the [VariableDeclaration]) which are
+ * considered.
+ * - [nodesToOp]: A mapping between CPG nodes and their operators used by the respective edges in
```
Comment (oxisto) [inline cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt:44]: We should consider renaming this to something more generic, Any other "user" of this class except Codyze will probably not understand what this map needs to contain.

File: cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+        }
+
+        return null
+    }
+
+    private fun Node.addEogPath(path: String) {
+        nodeIDtoEOGPathSet.computeIfAbsent(this.id) { mutableSetOf() }.add(path)
+    }
+
+    private fun Node.getEogPaths(): Set<String>? {
```
Comment (oxisto) [inline cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt:237]: nit picking, but `getXX` methods should be avoided in Kotlin. a read-only property should probably be used instead.

File: cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+        return nodeIDtoEOGPathSet[this.id]
+    }
+
+    /**
+     * If it's not an object-oriented language, we need to retrieve the base in a different way.
+     * Usually, it's the argument at index 0 but here, it's more obscure.
+     */
+    private fun Node.getThisPosition(): Int {
```
Comment (oxisto) [inline cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt:245]: see above.

File: cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+        return thisPositionOfNode[this]!!
```
Comment (oxisto) [inline cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt:246]: Avoid `!!`, instead an `Int?` should probably be returned which would also make `hasThisPosition` obsolete

File: cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt (hunk 6 above)
Comment (KuechA) [reply cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt:44]: Changed the name of the map to `nodeToRelevantMethod`. Is this easier to understand?

File: cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFA.kt (hunk 3 above)
Comment (KuechA) [reply cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFA.kt]: Changed the method name to `toDotString()`. In my opinion, it's easier to check if ones automaton is constructed correctly with such a graphical representation.

**Pull request merged**